File: node/generate.js
const fs = require('fs');
const fetch = require('node-fetch');
const path = require('path');
const pl = require('nodejs-polars');

const API_ROOT = 'https://community-api.coinmetrics.io/v4';
const MIN_FETCH_INTERVAL = 600; // rate limit interval in ms (used only for debugging)
const PAGE_SIZE = 10000;

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

async function fetchDataWithRetry(url) {
  let response;
  for (let attempt = 1; 1; attempt++) {
    try {
      response = await fetch(API_ROOT + url);
      if (response.ok) {
        const jsonResponse = await response.json();
        return jsonResponse.data;
      }
	  console.warn(`Fetch attempt ${attempt} failed due to HTTP: ${response.status} ${response.statusText}. See cf-ray: ${response.headers.get('cf-ray')} Retrying...`);
    } catch (error) {
        console.warn(`Fetch attempt ${attempt}: ${error}`);
    }
	await sleep(1000);
  }
}

const emitInfo = msg => {
	console.log(`[INF ${timestamp()}] ${msg}`);
};

const emitVerbose = msg => {
	if (process.env.VERBOSE) {
		emitInfo(msg);
	}
};

const emitErrorAndDie = msg => {
	console.error(`[ERR ${timestamp()}] ${msg} (status: ${lastStatus}`);
	process.exit(1);
};

let lastStatus = null;
const apiFetch = async path => fetch(API_ROOT + path)
	.then(res => {
		lastStatus = res.status;
		return res;
	})
	.then(res => res.json())
	.then(res => res.data);

const fsWrite = (filename, content) => {
	let ticker = filename;
	let oldFilename = path.resolve("../csv", `${ticker}.csv`);
	let oldFrame = pl.readCSV(oldFilename);
	let cols = oldFrame.columns;
	let dtypes = oldFrame.dtypes;
	let dtypeMap = {}
	for (let i = 0; i < cols.length; i++) {
		dtypeMap[cols[i]] = dtypes[i];
	}
	let newFrame = pl.readCSV(content, {dtypes: dtypeMap});
	newFrame.with

	let filteredOldFrame = oldFrame.filter(pl.col("time").lt(pl.col("time").max()));
	let filteredNewFrame = newFrame.join(filteredOldFrame, {on: "time", how:"anti"});
	let outFrame = pl.concat([filteredOldFrame, filteredNewFrame], {how: "diagonal"});

	filename = path.resolve(process.env.OUT, `${ticker}.csv`);
	// fs.writeFileSync(filename, content);
	outFrame.writeCSV(filename)
	emitVerbose(`Written ~${content.length} bytes to ${filename}`);
};

const metricIdsFromAssetInfo = assetInfo => assetInfo.metrics
	.filter(metric => metric.frequencies.find(freq => freq.frequency === '1d'))
	.map(metric => metric.metric);

const dataPointToCsvCols = (dataPoint, metricIds) => metricIds.map(metricId => dataPoint[metricId] ?? '').join(',');

const timestamp = () => new Date().toISOString();

const wait = interval => new Promise(res => setTimeout(res, interval));

const shouldOmitAsset = assetInfo => {
	return assetInfo.metrics == null || assetInfo.metrics.every(metric => metric.metric.startsWith('ReferenceRate'));
};

(async () => {
	const assetIndex = {};
	const assetIds = [];

	// build an asset info index and a list of asset IDs
	const assetInfo = await apiFetch('/catalog/assets');
	if (assetInfo == null) {
		emitErrorAndDie('Failed to fetch asset info');
		process.exit(1);
	}
	for (const info of assetInfo) {
		if (shouldOmitAsset(info)) {
			emitVerbose(`Omitting ${info.asset}`);
			continue;
		}
		assetIndex[info.asset] = info;
		assetIds.push(info.asset);
	}
	assetIds.sort();

	// retrieve metrics for every asset

	const totalAssets = assetIds.length;
	let fetchedAssets = 0;
	let lastFetchTime = 0;

	for (const assetId of (process.env.ASSETS?.split(',') ?? assetIds)) {
		const timeSinceLastFetch = Date.now() - lastFetchTime;
		if (process.env.THROTTLE && timeSinceLastFetch < MIN_FETCH_INTERVAL) {
			emitVerbose(`Pausing for ${MIN_FETCH_INTERVAL - timeSinceLastFetch}ms`);
			await wait(MIN_FETCH_INTERVAL - timeSinceLastFetch);
		}
		fetchedAssets++;
		const info = assetIndex[assetId];
		const metricIds = metricIdsFromAssetInfo(info);

		// fetch data
		emitInfo(`Fetching ${assetId} data with ${metricIds.length} metrics... (${fetchedAssets}/${totalAssets})`);
		lastFetchTime = Date.now();
		if (metricIds.length === 0) {
		    emitInfo(`Skipping asset: ${assetId} because there is no available metrics`)
		    continue
		}
		const seriesData = await fetchDataWithRetry(`/timeseries/asset-metrics/?assets=${assetId}&metrics=${encodeURIComponent(metricIds.join(','))}&page_size=${PAGE_SIZE}`);

		if (seriesData == null) {
			emitErrorAndDie(`Failed to fetch data for ${assetId}`);
		}

		// build CSV
		let csv = `time,${metricIds.join(',')}\n`;
		for (let dataPoint of seriesData) {
			csv += `${dataPoint.time.split('T')[0]},${dataPointToCsvCols(dataPoint, metricIds)}\n`;
		}

		// write to file
		fsWrite(assetId, csv);
		await sleep(750)
	}

	emitInfo('Finished');
})();

File: python/generate.py
from datetime import datetime

import pandas as pd
import asyncio
import httpx
import aiometer

from classes import CatalogV2AssetMetricsInfo, CatalogV2AssetMetricsResponse

API_ROOT = 'https://community-api.coinmetrics.io/v4'
PAGE_SIZE = 10000


async def fetch_data_with_retry(client: httpx.AsyncClient, url: str):
    attempt = 0
    while attempt < 5:
        try:
            resp = await client.get(url, timeout=10)
            # print(resp.text)
            resp.raise_for_status()
            return resp.json()
        except Exception as e:
            print(f"There was an error fetching data from {url}: {e}")
            attempt += 1
            await asyncio.sleep(1)


async def api_fetch(client: httpx.AsyncClient, path: str):
    try:
        resp = await client.get(f'{API_ROOT}/{path}')
        resp.raise_for_status()
        return resp.json()
    except httpx.HTTPError as e:
        print(f"HTTP Exception for {e.request.url} - {path}")
    pass


def metric_ids_from_asset_info(asset_info: CatalogV2AssetMetricsInfo) -> set[str]:
    return {
        m.metric
        for m in asset_info.metrics
        if [f for f in m.frequencies if f.frequency == "1d"]
    }

# @star_args
async def query_api_and_write_csv(args):
    total, client, i, metrics, assets = args
    print(f"Fetching {i}/{total} - {datetime.today()} - {assets}")
    url = API_ROOT + f'/timeseries/asset-metrics?page_size={PAGE_SIZE}&metrics={metrics}&assets={assets}&start_time=2024-12-05'
    resp = await fetch_data_with_retry(client, url)
    out_data = resp['data']
    next_page_url = resp.get('next_page_url')
    page = 1
    while next_page_url is not None:
        page += 1
        print(f"....{i}/{total}: page {page}....")
        resp = await fetch_data_with_retry(client, next_page_url)
        out_data.extend(resp['data'])
        next_page_url = resp.get('next_page_url')
    pd.DataFrame(out_data).to_csv(f"csv3/{i}.csv", index=False)


async def main():
    async with httpx.AsyncClient() as client:
        asset_metrics_resp = await api_fetch(client, '/catalog-v2/asset-metrics?page_size=10000')
        await asyncio.sleep(.6)
        asset_metrics = CatalogV2AssetMetricsResponse.from_dict(asset_metrics_resp)
        am_map = asset_metrics.combined_assets_and_metrics()

        total = len(am_map)
        it = [(total, client, i + 1, m, a) for i, (m, a) in enumerate(am_map.items())]
        await aiometer.run_on_each(query_api_and_write_csv, it, max_per_second=.6)


if __name__ == '__main__':
    asyncio.run(main())
